lxd/db/projects.go

```diff
@@ -26,3 +29,131 @@ func (db *DB) GetProject(ctx context.Context, projectName string) (*cluster.Proj
 
 	return p, nil
 }
+
+// If the project has the profiles feature enabled, we use its own
+// profiles to expand the instances configs, otherwise we use the
+// profiles from the default project.
+func projectInheritsProfiles(project *api.Project) bool {
+	return shared.IsFalseOrEmpty(project.Config["features.profiles"])
+}
+
+// ProjectArgs is all the instances and profiles associated with a project.
+// If Project.Config["features.profiles"] is false, Profiles is the default
+// project's profiles.
+type ProjectArgs struct {
+	Profiles  []api.Profile
+	Instances []api.Instance
+}
+
+// GetProjectInstancesAndProfiles gets all instances and profiles associated
+// with some projects in as few queries as possible.
+// Returns a map[projectName]args.
+func (c *ClusterTx) GetProjectInstancesAndProfiles(ctx context.Context, projects ...*api.Project) (map[string]*ProjectArgs, error) {
```

Comment:
As maps are returned by reference, is there a reason the map stores a *ProjectArgs and not just ProjectArgs?

Reply:
Not a great one; it allows this:
```go
// line 106
projectArgs[profile.Project].Profiles = append(projectArgs[profile.Project].Profiles, *apiProfile)
```
as opposed to
``` go
args := projectArgs[profile.Project]
args.Profiles = append(args.Profiles, *apiProfile)
projectArgs[profile.Project] = args
```
since map keys aren't addressable. Not sure if the compiler would optimize that or not; I only really care about it being a bit more concise.